### PR TITLE
add a few type declarations to REPL

### PR DESCRIPTION
**Pull request by KristofferC**
Removes ~150 invalidations from `lastindex` (when e.g. loading `LaTeXStrings`).

### CODE REVIEW

**Review by timholy — APPROVED** [review]
Would be really nice to have these keymap methods documented...this was the hardest part of REPL for me to figure out what argument types can be counted on, and as you noticed here I just ended up punting on some methods. Which you then discovered still needed to be tackled after all.

Any words of wisdom from @Keno would of course be appreciated, but this seems OK to me.

File: stdlib/REPL/src/LineEdit.jl [hunk 1]
```diff
@@ -1577,7 +1578,7 @@ function keymap_merge(target::Dict{Char,Any}, source::Union{Dict{Char,Any},AnyDi
         if isa(value, Union{Char,String})
             value = getEntry(ret, value)
             if value === nothing
-                error("Could not find redirected value " * escape_string(source[key]))
+                error("Could not find redirected value " * escape_string(String(source[key])))
```
Comment (KristofferC) [inline stdlib/REPL/src/LineEdit.jl:1581]: Actually, this doesn't work, I think it hits the no inference on error path limitation.

Comment (timholy) [reply stdlib/REPL/src/LineEdit.jl:1581]: How about
```julia
newvalue = getEntry(ret, value)
if newvalue === nothing
    error("Could not find redirected value " * escape_string(String(value)))
end
value = newvalue
```

File: stdlib/REPL/src/LineEdit.jl [hunk 2]
```diff
@@ -1577,14 +1578,19 @@ function keymap_merge(target::Dict{Char,Any}, source::Union{Dict{Char,Any},AnyDi
...
         end
         add_nested_key!(ret, key, value; override = true)
     end
     return ret
 end
 
+throw_eager_redirection_cycle(key::Union{Char, String}) =
+    error("Eager redirection cycle detected for key ", escape_string(key))
+throw_could_not_find_redirected_value(key::Union{Char, String}) =
+    error("Could not find redirected value ", escape_string(key))
```
Comment (vtjnash) [inline stdlib/REPL/src/LineEdit.jl:1592]:
```suggestion
    error("Eager redirection cycle detected for key ", repr(key))
throw_could_not_find_redirected_value(key::Union{Char, String}) =
    error("Could not find redirected value ", repr(key))
```
nit: should be the more standard name for this

**Comment by KristofferC** [comment]
> Would be really nice to have these keymap methods documented...

Well, it's nice that they are not documented because then they are internal and can be changed ;)

**Pull request merged**